File: 0849-maximize-distance-to-closest-person/0849-maximize-distance-to-closest-person.java
class Solution {
    public int maxDistToClosest(int[] seats) {
        int n = seats.length;
        int[] dis = new int[n];
        
        Arrays.fill(dis, 20001);
        if(seats[n-1]==1) dis[n-1] = 0;
        
        for(int i=n-2; i>=0; --i){
            dis[i] = (seats[i]==1) ? 0 : dis[i+1]+1;
        }
        
        int maxdis = Integer.MIN_VALUE, curr = 20001;
        for(int i=0; i<n; ++i){
            curr = (seats[i]==1) ? 0 : curr+1;
            maxdis = Math.max(maxdis, Math.min(curr, dis[i]));
        }
        
        return maxdis;
    }
}

File: 0849-maximize-distance-to-closest-person/0849-maximize-distance-to-closest-person.py
class Solution:
    def maxDistToClosest(self, seats: List[int]) -> int:
        n = len(seats)
        dis = [float('inf') for i in seats]
        for i in range(n-1, -1, -1):
            if i==n-1 and seats[i]:
                dis[i] = 0
            elif i<n-1:
                dis[i] = 0 if seats[i] else dis[i+1]+1
        
        ans = -float('inf')
        curr = float('inf')
        for i in range(n):
            curr = 0 if seats[i] else curr+1
            ans = max(ans, min(curr, dis[i]))
        return ans
            
            